### PR TITLE
V2 revisit modifiers

### LINKED ISSUE #696: Non-manifold vertices when extruding SVG or subtracting shapes
## Expected Behavior

Extruding an SVG and subtracting a cylinder from a cube should result in manifold models. See SCAD code provided.

## Actual Behavior

Extruding an SVG creates a model with non-manifold vertices. Same thing when subtracting a 2 cylinders from a cube. See JSCAD code provided. This leads to errors in PrusaSlicer.

## Steps to Reproduce the Problem

The STL produced here is shown as having non-manifold vertices in meshlab.

```js
const fs = require('fs').promises

const { geometries, primitives, booleans, extrusions, transforms } = require('../OpenJSCAD.org/packages/modeling')
const { svgDeSerializer, stlSerializer } = require('../OpenJSCAD.org/packages/io/io')

async function cast () {
	const height = 25.4
	const width = 12
	const depth = 22
	const scale = height/720
	const extrudeHeight = 1.32

	const shapes = svgDeSerializer.deserialize({
		output: 'geometry',
		target: 'path', 
		segments: 100
	}, '<svg PXPMM="1" xmlns="http://www.w3.org/2000/svg"><path d="M258 108C280 168 301 260 314 403L258 403C252 338 243 259 228 194C210 255 190 330 176 403C222 435 268 471 268 575C268 703 199 707 184 707C145 707 94 674 94 557C94 522 99 473 107 419C63 391 22 354 22 238C22 97 75-2 151-2C167-2 194 1 223 39C240 13 259-3 279-3C288-3 299-2 306 1L311 54C284 53 268 84 258 108ZM155 582C155 629 162 651 180 651C202 651 210 617 210 575C210 512 192 480 166 458C159 503 155 545 155 582ZM154 65C102 65 85 193 85 249C85 299 98 330 118 354C136 262 162 164 193 94C182 76 169 65 154 65Z" /></svg>')

	const poly = booleans.union(shapes.map((shape) => primitives.polygon({ points: shape.points })))
	const polyLetter = transforms.scale([scale, scale, 1], poly)
	let letterShape = extrusions.extrudeLinear({ height: extrudeHeight }, geometries.geom2.reverse(polyLetter))

	const body = booleans.subtract(
		booleans.union(
			primitives.cuboid({ 
				size: [width, height, depth],
				center: [0, 0, 0]
			}),
			transforms.translateZ(depth/2 + extrudeHeight/2, transforms.center({}, transforms.mirrorX(transforms.mirrorY(letterShape)))),
		),
		transforms.translate([-2, height/2, 0], transforms.rotateY(1.5708, primitives.cylinder({ 
			height: width,
			radius: depth * 0.4,
			segments: 100
		}))),
		transforms.translate([-2, -height/2, 0], transforms.rotateY(1.5708, primitives.cylinder({ 
			height: width,
			radius: depth * 0.4,
			segments: 100
		}))) 
	)

	const stl = Buffer.concat(stlSerializer.serialize({ binary: true }, body).map((buff) => Buffer.from(buff)))

	await fs.writeFile('./test-manifold.stl', stl, 'binary')
}

cast()
```

Save this code as the test SVG file:

```
<svg PXPMM="1" xmlns="http://www.w3.org/2000/svg"><path d="M258 108C280 168 301 260 314 403L258 403C252 338 243 259 228 194C210 255 190 330 176 403C222 435 268 471 268 575C268 703 199 707 184 707C145 707 94 674 94 557C94 522 99 473 107 419C63 391 22 354 22 238C22 97 75-2 151-2C167-2 194 1 223 39C240 13 259-3 279-3C288-3 299-2 306 1L311 54C284 53 268 84 258 108ZM155 582C155 629 162 651 180 651C202 651 210 617 210 575C210 512 192 480 166 458C159 503 155 545 155 582ZM154 65C102 65 85 193 85 249C85 299 98 330 118 354C136 262 162 164 193 94C182 76 169 65 154 65Z" /></svg>```


This is an equivalent SCAD code which produces a similar model (I couldn't find the resize method in JSCAD) which is shown as having 0 non-manifold vertices or edges in meshlab.

```
height=25.4;
width=12;
depth=22;

difference() {
    union() {
        scale([-1, -1, 1]) {
            linear_extrude(height=1.32) {
                resize([0, height, 0], auto=true) { 
                    import("./test.svg", center=true);
                }
            }
        }

        translate([-12/2, -height/2, -depth]) {
            difference() {
                cube([width, height, depth]);
            }
        }
    }
 
    translate([4, -height/2, -depth/2]) {
        rotate([0,-90,0]) {
            cylinder(width, r=depth*0.4, $fn=100);
        }
    }
 
    translate([4, height/2, -depth/2]) {
        rotate([0,-90,0]) {
            cylinder(width, r=depth*0.4, $fn=100);
        }
    }
 }
```

## Specifications

  - Version: 2.0.0-alpha2
  - Platform: macOS 10.14.6
  - Environment: Node.js v12.13.0

### LINKED ISSUE #758: V2 extrusion issues
## Expected Behavior

There seems to be some issues with the extrusion logic.
- Only closed slices (those that defined closed sets of edges) can be extruded
- Uncapped extrusions seem strange, where sides are built between end edge and start edge

## Steps to Reproduce the Problem

  1.
  2.
  3.

## Specifications

  - Version: V2
  - Platform: All
  - Environment: (browser, local server, node.js etc) ALL

### PR DESCRIPTION
**Pull request by z3dev**
This is the second try at converting meshes to triangles for exporting to external formats. There were some changes to the modifiers:
- added insertTjunctions() to place addition vertices at the T junctions (converted and improved V1 code)
- enhanced the triangulatePolygons() to be smarter when creating triangles for LARGE polygons (those with T junctions)
- adjusted generalize() to use the above properly
- adjusted tests

In addition to the above, the following IO serializers have been updated to use the generalize() function, as these external formats require triangles.
- amf-serializer
- dxf-serializer
- stl-serializer
- x3d-serializer

Fixes #696 
Fixes #758 

### All Submissions:

* [X] Have you followed the guidelines in our Contributing document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [X] Does your submission pass tests?

### CODE REVIEW

**Comment by z3dev** [comment]
@SimonClark please review and let me know if you find anything inconsistent or wrong.

I think this is ready to be merged now. I tested everything that I could, including exports / imports.

File: packages/modeling/src/operations/modifiers/comparePolygons.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+const comparePolygon = require('../../../test/helpers/comparePolygons') // rename
```
Comment (SimonClark) [inline packages/modeling/src/operations/modifiers/comparePolygons.js:1]: Is this production code relying on test code?

File: packages/modeling/src/operations/modifiers/insertTjunctions.js [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+  Note this function is meant to be used to preprocess geometries when triangulation is required, i.e. AMF, STL, etc.
+  Do not use the results in other operations.
+*/
+
+/*
+ * Insert missing vertices for T junctions, which creates polygons that can be triangulated.
+ * @param {Array} polygons - the original polygons which may or may not have T junctions
+ * @return original polygons (if no T junctions found) or new polygons with updated vertices
+ */
+const insertTjunctions = (polygons) => {
```
Comment (SimonClark) [inline packages/modeling/src/operations/modifiers/insertTjunctions.js:112]: This method is a bit enormous. Are there any easily extractable chunks we could pull out to make it more readable/maintainable?

File: packages/modeling/src/operations/modifiers/comparePolygons.js (hunk 1 above)
Comment (z3dev) [reply packages/modeling/src/operations/modifiers/comparePolygons.js]: UG. I forgot. I renamed and moved to test/helpers.

DONE

File: packages/modeling/src/operations/modifiers/insertTjunctions.js (hunk 2 above)
Comment (z3dev) [reply packages/modeling/src/operations/modifiers/insertTjunctions.js:112]: Actually, there are three sets of 'edge' implementations in these functions.
- edges.js
- edges in insertTjunctions.js
- edges in mergePolygons.js

A single implement would be ideal but I'm not really motivated. I'm kind of tired of modifiers.

Comment (z3dev) [reply packages/modeling/src/operations/modifiers/insertTjunctions.js:112]: Oh. And there's another in the quickhull piece as well.

Comment (z3dev) [reply packages/modeling/src/operations/modifiers/insertTjunctions.js:112]: @SimonClark thoughts? how about a new issue to revisit this later?

Comment (SimonClark) [reply packages/modeling/src/operations/modifiers/insertTjunctions.js:112]: sounds fine

**Review by SimonClark — APPROVED** [review]

**Pull request merged**